Implement kafka-streams-groups.sh --list

aliehsaeedii opened the pull request:
Implement the core of `kafka-streams-groups.sh`
Implement `--list` and its options: (only `--state`)

aliehsaeedii committed:
impl --list

aliehsaeedii commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java:
```diff
@@ -0,0 +1,170 @@
+        try {
+            opts.checkArgs();
+
+            CommandLineUtils.maybePrintHelpOrVersion(opts, "This tool helps to list all streams groups.");
```
The message should change as new command are added.

aliehsaeedii commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java:
```diff
@@ -0,0 +1,170 @@
+            // should have exactly one action
+            long actions = Stream.of(opts.listOpt).filter(opts.options::has).count();
+            if (actions != 1)
+                CommandLineUtils.printUsageAndExit(opts.parser, "Command must include exactly one action: --list.");
```
The same here.

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+
+public class StreamsGroupCommandOptions extends CommandDefaultOptions {
+    public static final String BOOTSTRAP_SERVER_DOC = "REQUIRED: The server(s) to connect to.";
+    public static final String LIST_DOC = "List all share groups.";
```
"List all streams groups"

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+
+import joptsimple.OptionSpec;
+
+
```
double new line

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String TIMEOUT_MS_DOC = "The timeout that can be set for some use cases. For example, it can be used when describing the group " +
+        "to specify the maximum amount of time in milliseconds to wait before the group stabilizes.";
+    public static final String COMMAND_CONFIG_DOC = "Property file containing configs to be passed to Admin Client.";
+    public static final String DRY_RUN_DOC = "Only show results without executing changes on share groups. Supported operations: reset-offsets.";
```
We do not need --dry-run yet, as it's only supported by reset-offsets. Also, this shouldn't mention share groups.

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+        "to specify the maximum amount of time in milliseconds to wait before the group stabilizes.";
+    public static final String COMMAND_CONFIG_DOC = "Property file containing configs to be passed to Admin Client.";
+    public static final String DRY_RUN_DOC = "Only show results without executing changes on share groups. Supported operations: reset-offsets.";
+    public static final String EXECUTE_DOC = "Execute operation. Supported operations: reset-offsets.";
```
remove

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String COMMAND_CONFIG_DOC = "Property file containing configs to be passed to Admin Client.";
+    public static final String DRY_RUN_DOC = "Only show results without executing changes on share groups. Supported operations: reset-offsets.";
+    public static final String EXECUTE_DOC = "Execute operation. Supported operations: reset-offsets.";
+    public static final String STATE_DOC = "When specified with '--describe', includes the state of the group." + NL +
```
Seems like this shouldn't mention `--describe` yet.

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String EXECUTE_DOC = "Execute operation. Supported operations: reset-offsets.";
+    public static final String STATE_DOC = "When specified with '--describe', includes the state of the group." + NL +
+        "When specified with '--list', it displays the state of all groups. It can also be used to list groups with specific states. " +
+        "Valid values are Empty, Stable and Dead.";
```
There are more possible states than just `Empty, Stable and Dead`.

lucasbru commented on tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void checkArgs() {
+        CommandLineUtils.maybePrintHelpOrVersion(this, "This tool helps to list, describe, reset and delete share groups.");
```
Mentions share groups.
reset and delete and describe are not probided yet

lucasbru commented on tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandTest.java:
```diff
@@ -0,0 +1,177 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class StreamsGroupCommandTest {
```
We'll also want to have an integration test for the tool as well, similar to `ListConsumerGroupTest`, which actually runs a cluster. Maybe we could rename this to `StreamsGroupCommandUnitTest` and add a `StreamsGroupCommandTest` that uses a similar pattern as `ListConsumerGroupTest`?

lucasbru reviewed (approved):
Thanks for the PR! I left some comments

lucasbru reviewed (commented):
Thanks for the PR! I left some comments

aliehsaeedii committed:
address reviews

aliehsaeedii commented:
> Thanks for the PR! I left some comments

Thanks, @lucasbru. I addressed the nits. I think we can merge this PR and meanwhile I work on the integration test.

aliehsaeedii committed:
uncomment type

lucasbru commented on tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandUnitTest.java:
```diff
@@ -0,0 +1,177 @@
+
+
+        result = StreamsGroupCommand.groupStatesFromString("PREPARINGREBALANCE");
+        assertEquals(new HashSet<>(Collections.singletonList(GroupState.PREPARING_REBALANCE)), result);
```
@aliehsaeedii as per discussion, can you please remove preparing_rebalance and completing rebalance from the possible states in `GroupState` for streams, and this test? Thanks!

aliehsaeedii committed:
remove wrong states

aliehsaeedii committed:
Merge remote-tracking branch 'origin/add_kstreams_groups_sh' into add_kstreams_groups_sh

lucasbru reviewed (approved):
Thanks for the PR! LGTM
